client/src/auth/PrivateRoute.jsx

```diff
@@ -0,0 +1,71 @@
+import React, { Component } from "react";
+import { Route, withRouter } from "react-router-dom";
+import axios from 'axios';
+
+class PrivateRoute extends Component {
+  initialMount = false;
+
+  // axios configuration
+  axiosSource = axios.CancelToken.source();
+  axiosInstance = axios.create({
+    baseURL: '/api/v1/',
+    cancelToken: this.axiosSource.token
+  });
+
+  state = {
+    isAuth: false,
+  }
+
+  componentWillUnmount() {
+    this.axiosSource.cancel('cancel isAuthenticated request');
+  }
+
+  componentDidMount() {
+    this.axiosInstance
+    .get('/isAuthenticated')
+      .then(() => {
+        this.initialMount = true;
+        this.setState({ isAuth: true });
+      })
+      .catch(() => {
+        this.isUnAuthorized();
+      });
+  }
+
+  componentDidUpdate() {
+    // this will stop when updating after the initial mount
+    if (!this.initialMount) {
+      this.axiosInstance
+      .get('/isAuthenticated')
+        .then(() => {
+          if (!this.state.isAuth) {
+            this.setState({ isAuth: true });
+          }
+        })
+        .catch(() => {
+          this.isUnAuthorized();
+        });
+    } else {
+      this.initialMount = false;
+    }
+  }
+  
+  isUnAuthorized = () => {
+    this.props.history.push('/login', { from: this.props.location.pathname});
+  }
+
+  render() {
+    const { component: Component, ...rest } = this.props;
+    const { isAuth } = this.state;
+
+    return (
+      isAuth &&
+      <Route
```

Comment:
@ashatat I wonder if you need to always render the route, and then conditionally return null inside the route's render.  In other words, this:
```
return (
  <Route
    {...rest}
    render={props => isAuth && <Component {...props}/>}
  />
);
```

My worry is that if isAuth is false because the authentication request has not yet finished, then having no `<Route` rendered at all will break the path-matching and your app's `<Switch` will match some other route instead (essentially redirecting the user somewhere else if the authentication request is slow.  What do you think?

Reply:
This has been troubling me,
- <PrivateRoute /> is just a wrapper for the <Route /> component, yet when I go to a route that is not matched I get the `PageNotFound` component.
- I tried to Throttling the network to slow 3G which made the request to the API take some time but the `PageNotFound` component doesn't get rendered, nothing gets rendered until the request come back and set the isAuth to true then it renders the component
- I add console.log inside the render function of both  <PrivateRoute> and <LogedoutRoute>, and tried a not matched route and it renders <PageNotFound/> without getting into the render function of both Private and Logedout Route

**what you said make sense and I will always put the condition inside the render of the Route  but I am really confused how the PrivateRoute worked as Route** :sob:

Reply:
I'm confused too now 😅
I looked at the Switch source ([here](https://github.com/ReactTraining/react-router/blob/master/packages/react-router/modules/Switch.js)) and it looks like it iterates over its children and just checks the child instance's props directly, when deciding which element matches and should render.  So I guess that means as long as the wrapper component has a `path` prop, it treats it like it's a `<Route />` even if the wrapped component isn't currently rendering a Route, or never renders a Route at all.  I guess that makes sense because otherwise the router code would be stuck, not knowing if it can render a route because it can't check that the route matches without first rendering it 🤯
So I wonder if that means this code would work even if it didn't render a <Route /> at all, and just rendered the component?!
Anyways, I think that means that this code is fine as-is.  Haha.  Sorry for the distraction!

Reply:
That's really interesting 
I removed <Route > from the PrivetRoute and it worked just fine :smile: 
```jsx
return isAuth && <Component {...rest} />
```
I peeked inside the Switch source code, I understood NO THING :rofl:, but can you get the props of a child component? :open_mouth: 


Really Thanks for the information, it's not a distraction at all, we love to learn new stuff
LOL should I remove <Route /> from the <PrivetRoute />  ? :rofl:

Reply:
Haha, it’s up to you if you want to change it or not - I can’t think of any reasons why one way is better than the other, after all we have learned. 
I didnt know that you could get the props of a child component either, but I do know that you can get the children of a component by using this.props.children (it’s just an array of components) which is pretty useful.